users/api/groups.go

```diff
@@ -28,121 +30,137 @@ func groupsHandler(svc groups.Service, authClient auth.AuthClient, r *chi.Mux, l
 		kithttp.ServerErrorEncoder(apiutil.LoggingErrorEncoder(logger, api.EncodeError)),
 	}
 
-	r.Route("/groups", func(r chi.Router) {
-		r.Post("/", otelhttp.NewHandler(kithttp.NewServer(
-			gapi.CreateGroupEndpoint(svc, authClient, policies.NewGroupKind),
-			gapi.DecodeGroupCreate,
-			api.EncodeResponse,
-			opts...,
-		), "create_group").ServeHTTP)
-
-		r.Get("/{groupID}", otelhttp.NewHandler(kithttp.NewServer(
-			gapi.ViewGroupEndpoint(svc, authClient),
-			gapi.DecodeGroupRequest,
-			api.EncodeResponse,
-			opts...,
-		), "view_group").ServeHTTP)
-
-		r.Delete("/{groupID}", otelhttp.NewHandler(kithttp.NewServer(
-			gapi.DeleteGroupEndpoint(svc, authClient),
-			gapi.DecodeGroupRequest,
-			api.EncodeResponse,
-			opts...,
-		), "delete_group").ServeHTTP)
-
-		r.Get("/{groupID}/permissions", otelhttp.NewHandler(kithttp.NewServer(
-			gapi.ViewGroupPermsEndpoint(svc, authClient),
-			gapi.DecodeGroupPermsRequest,
-			api.EncodeResponse,
-			opts...,
-		), "view_group_permissions").ServeHTTP)
-
-		r.Put("/{groupID}", otelhttp.NewHandler(kithttp.NewServer(
-			gapi.UpdateGroupEndpoint(svc, authClient),
-			gapi.DecodeGroupUpdate,
-			api.EncodeResponse,
-			opts...,
-		), "update_group").ServeHTTP)
-
-		r.Get("/", otelhttp.NewHandler(kithttp.NewServer(
-			gapi.ListGroupsEndpoint(svc, authClient, "groups", "users"),
+	checkSuperAdminMiddleware := api.CheckSuperAdminMiddleware(authClient)
+	r.Group(func(r chi.Router) {
+		r.Use(api.IdentifyMiddleware(authClient))
+
+		r.Route("/groups", func(r chi.Router) {
+			authzMiddleware := api.AuthorizeMiddleware(authClient, gapi.CreateGroupAuthReq)
```

Comment:
Why don't we have Authorization as Service middleware, Instead of having In API?

Reply:
Since any how for each service function, we need to have its own authorization function. 
So having in Service middleware like logging , metering will make ensure that all the function have authorization function.

Reply:
Because that will alter the service interface. I.e. for auth, we need only a token; for other service functions - we need a session (user ID, domain ID, object ID, whatever), so we would have to use them exclusively (token only for auth middleware, session for everything else) while keeping them both in the interface - resulting in the poor design.